core/jvm/src/main/scala/scalapb/zio_grpc/ZCall.scala

```diff
@@ -0,0 +1,39 @@
+package scalapb.zio_grpc
+
+import io.grpc.Status
+import scalapb.zio_grpc.ZCall.ReadyPromise
+import zio.{Promise, Ref, UIO, ZIO}
+
+object ZCall {
+  type ReadyPromise = Ref[Option[Promise[Nothing, Unit]]]
+}
+
+trait ZCall[-R, A] extends Any {
+
+  def request(n: Int): ZIO[R, Status, Unit]
+
+  def sendMessage(message: A): ZIO[R, Status, Unit]
+
+  def sendMessageWhenReady(message: A): ZIO[R, Status, Unit] =
+    ZIO.ifM(isReady)(
+      sendMessage(message),
+      Promise.make[Nothing, Unit].flatMap { promise =>
+        readyPromise.set(Some(promise)) *> ZIO.ifM(isReady)(
+          sendMessage(message),
+          promise.await *> sendMessage(message)
+        )
+      }
+    )
+
+  def onReady(): ZIO[R, Status, Unit] =
+    readyPromise
+      .modify[UIO[Unit]] {
+        case Some(promise) => promise.succeed(()).unit -> None
+        case None          => ZIO.unit                 -> None
+      }
+      .flatten
```

Comment:
Just realised the recommendation to check `isReady` in `onReady` handler:

> Because there is a processing delay to deliver this notification, it is possible for concurrent writes to cause isReady() == false within this callback. Handle "spurious" notifications by checking isReady()'s current value instead of assuming it is now true. If isReady() == false the normal expectations apply, so there would be another onReady() callback.  

Should I add this check?

```suggestion
      .flatten.whenM(isReady)
```

Reply:
I think it makes sense to follow the api doc recommendation.